File: app/src/main/java/com/example/beto/viagemplanejada/ViagemAdapter.java
package com.example.beto.viagemplanejada;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

public class ViagemAdapter extends RecyclerView.Adapter {

    List<DataSnapshot> items;
    private Context context;

    public ViagemAdapter(List<DataSnapshot> viagens) {
        items = viagens;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View itemView = LayoutInflater.from(parent.getContext()).
                inflate(R.layout.view_holder_layout, parent, false);
        context = parent.getContext();
        return new ViagemViewHolder(itemView);
    }


    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, final int position) {
        Publicacao publicacao = items.get(position).getValue(Publicacao.class);

        ViagemViewHolder viagemViewHolder = (ViagemViewHolder) holder;

        viagemViewHolder.paisTextView.setText(publicacao.getPais());

        Date dtConvert = publicacao.getDtViagem();
        DateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");

        viagemViewHolder.dtTextView.setText(dateFormat.format(dtConvert));
        viagemViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                String id = items.get(position).getKey();
                Intent intent = new Intent(context, AddPublicacao.class);
                intent.putExtra("id",id);
                context.startActivity(intent);
            }
        });




    }

    @Override
    public int getItemCount() {
        return items.size();
    }




    public void addItem(DataSnapshot novaPublicacao){
        items.add(novaPublicacao);

        notifyItemInserted(items.size()-1);
    }


    public void changeItem(DataSnapshot changed){
        for (int i = 0; i < items.size(); i++) {
            if (changed.getKey().equals(items.get(i).getKey())){
                items.set(i, changed);
                notifyItemChanged(i);
            }
        }
    }

    public void removeItem(DataSnapshot removed){
        for (int i = 0; i < items.size(); i++) {
            if (removed.getKey().equals(items.get(i).getKey())){
                items.remove(i);
                notifyItemRemoved(i);
            }
        }
    }

    public String recuperaId(int index){

        String id = items.get(index).getKey();
        return id;
    }
    public class ViagemViewHolder extends RecyclerView.ViewHolder{

        public TextView paisTextView;
        public TextView dtTextView;
        public ImageButton deleteButton;

        public ViagemViewHolder(@NonNull View itemView) {
            super(itemView);

            paisTextView = itemView.findViewById(R.id.lblPais);
            dtTextView = itemView.findViewById(R.id.txtDtViagem);
            deleteButton = itemView.findViewById(R.id.imgDeleteBtn);



            deleteButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    int index = getAdapterPosition();
                    String idToRemove = items.get(index).getKey();

                    DatabaseReference ref = FirebaseDatabase
                            .getInstance()
                            .getReference("Publicacao")
                            .child(idToRemove);
                    ref.removeValue();
                }
            });
        }
    }
}

File: app/src/main/java/com/example/beto/viagemplanejada/ViagemAdapter.kt
package com.example.beto.viagemplanejada

import android.content.Context
import android.content.Intent

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.beto.viagemplanejada.model.Publicacao

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.FirebaseDatabase

import java.text.SimpleDateFormat

class ViagemAdapter(internal var items: MutableList<DataSnapshot>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var context: Context? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.view_holder_layout, parent, false)
        context = parent.context
        return ViagemViewHolder(itemView)
    }


    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val publicacao = items[position].getValue<Publicacao>(Publicacao::class.java!!)

        val viagemViewHolder = holder as ViagemViewHolder

        viagemViewHolder.paisTextView.setText(publicacao!!.pais)

        val dtConvert = publicacao!!.dtViagem
        val dateFormat = SimpleDateFormat("dd/MM/yyyy")

        viagemViewHolder.dtTextView.text = dateFormat.format(dtConvert)
        viagemViewHolder.itemView.setOnClickListener {
            val id = items[position].key
            val intent = Intent(context, AddPublicacao::class.java)
            intent.putExtra("id", id)
            context!!.startActivity(intent)
        }


    }

    override fun getItemCount(): Int {
        return items.size
    }


    fun addItem(novaPublicacao: DataSnapshot) {
        items.add(novaPublicacao)

        notifyItemInserted(items.size - 1)
    }


    fun changeItem(changed: DataSnapshot) {
        for (i in items.indices) {
            if (changed.key == items[i].key) {
                items[i] = changed
                notifyItemChanged(i)
            }
        }
    }

    fun removeItem(removed: DataSnapshot) {
        for (i in items.indices) {
            if (removed.key == items[i].key) {
                items.removeAt(i)
                notifyItemRemoved(i)
            }
        }
    }

    inner class ViagemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var paisTextView: TextView
        var dtTextView: TextView
        var deleteButton: ImageButton

        init {

            paisTextView = itemView.findViewById(R.id.lblPais)
            dtTextView = itemView.findViewById(R.id.txtDtViagem)
            deleteButton = itemView.findViewById(R.id.imgDeleteBtn)



            deleteButton.setOnClickListener {
                val index = adapterPosition
                val idToRemove = items[index].key

                val ref = FirebaseDatabase
                        .getInstance()
                        .getReference("Publicacao")
                        .child(idToRemove!!)
                ref.removeValue()
            }
        }
    }
}
